File: backend/app/services/aspect_engine.py
"""Planetary aspect calculation engine for Vedic and Western astrology."""

from dataclasses import dataclass

from app.models.chart_models import PlanetPosition

# ── Western aspect definitions ───────────────────────────────
# Each entry maps an aspect name to its ideal angle and allowed orb.

WESTERN_ASPECTS: dict[str, dict[str, float]] = {
    "Conjunction": {"angle": 0.0, "orb": 8.0},
    "Opposition": {"angle": 180.0, "orb": 8.0},
    "Trine": {"angle": 120.0, "orb": 8.0},
    "Square": {"angle": 90.0, "orb": 7.0},
    "Sextile": {"angle": 60.0, "orb": 6.0},
}

# ── Vedic special (graha-drishti) aspects ────────────────────
# Mars, Jupiter, and Saturn cast special aspects to specific houses
# counted from their own position.  The house offset is converted to
# degrees by multiplying by 30.

VEDIC_SPECIAL_ASPECTS: dict[str, list[int]] = {
    "Mars": [4, 8],
    "Jupiter": [5, 9],
    "Saturn": [3, 10],
}

# ── Speed ordering (fastest → slowest) for applying logic ────
_PLANET_SPEED_ORDER: list[str] = [
    "Moon",
    "Mercury",
    "Venus",
    "Sun",
    "Mars",
    "Jupiter",
    "Saturn",
    "Rahu",
    "Ketu",
]


@dataclass
class AspectData:
    """Represents a single aspect between two planets."""

    planet1: str
    planet2: str
    aspect_type: str
    exact_angle: float
    orb: float
    applying: bool
    vedic: bool


class AspectEngine:
    """Calculate Western and Vedic planetary aspects."""

    # ── helpers ───────────────────────────────────────────────

    @staticmethod
    def _angle_between(lon1: float, lon2: float) -> float:
        """Return the angular distance between two ecliptic longitudes (0-180)."""
        diff = abs(lon1 - lon2) % 360.0
        return diff if diff <= 180.0 else 360.0 - diff

    @staticmethod
    def _is_applying(p1: PlanetPosition, p2: PlanetPosition, target_angle: float) -> bool:
        """Approximate whether the aspect is applying.

        A faster planet approaching the exact aspect angle with a slower
        planet is considered *applying*.  We use mean speed ordering
        rather than instantaneous speed for a simple, deterministic check.
        """
        speed_rank = {name: idx for idx, name in enumerate(_PLANET_SPEED_ORDER)}
        rank1 = speed_rank.get(p1.name, 99)
        rank2 = speed_rank.get(p2.name, 99)

        # The faster planet (lower rank number) is the one that "applies"
        # towards the slower.  If the faster planet still has arc to
        # cover before reaching the exact aspect angle, the aspect is
        # applying; otherwise it is separating.
        if rank1 < rank2:
            # p1 is faster
            faster, slower = p1, p2
        else:
            faster, slower = p2, p1

        current_angle = AspectEngine._angle_between(faster.longitude, slower.longitude)
        # If the current distance is slightly larger than the target the
        # faster body is still closing the gap → applying.
        return current_angle >= target_angle

    # ── main calculation ─────────────────────────────────────

    @classmethod
    def calculate_aspects(
        cls,
        planets: list[PlanetPosition],
        include_vedic: bool = True,
    ) -> list[AspectData]:
        """Return all aspects found among *planets*.

        Parameters
        ----------
        planets:
            List of planet positions (longitude, sign, etc.).
        include_vedic:
            When ``True`` (default), Vedic special aspects (graha-drishti)
            are included in addition to the five standard Western aspects.

        Returns
        -------
        list[AspectData]
            Aspects sorted by tightest orb (smallest first).
        """
        aspects: list[AspectData] = []
        # Track pairs already matched so Vedic pass can skip duplicates.
        seen_pairs: set[tuple[str, str]] = set()

        n = len(planets)

        # ── Western aspects ──────────────────────────────────
        for i in range(n):
            for j in range(i + 1, n):
                p1, p2 = planets[i], planets[j]
                angle = cls._angle_between(p1.longitude, p2.longitude)

                for aspect_name, spec in WESTERN_ASPECTS.items():
                    target = spec["angle"]
                    max_orb = spec["orb"]
                    orb = abs(angle - target)
                    if orb <= max_orb:
                        applying = cls._is_applying(p1, p2, target)
                        aspects.append(
                            AspectData(
                                planet1=p1.name,
                                planet2=p2.name,
                                aspect_type=aspect_name,
                                exact_angle=target,
                                orb=round(orb, 4),
                                applying=applying,
                                vedic=False,
                            )
                        )
                        pair_key = (
                            (min(p1.name, p2.name), max(p1.name, p2.name))
                        )
                        seen_pairs.add(pair_key)

        # ── Vedic special aspects (graha-drishti) ────────────
        if include_vedic:
            vedic_orb = 10.0  # degrees

            for caster in planets:
                house_offsets = VEDIC_SPECIAL_ASPECTS.get(caster.name)
                if house_offsets is None:
                    continue

                for offset in house_offsets:
                    target_lon = (caster.longitude + offset * 30.0) % 360.0

                    for receiver in planets:
                        if receiver.name == caster.name:
                            continue

                        pair_key = (
                            min(caster.name, receiver.name),
                            max(caster.name, receiver.name),
                        )
                        if pair_key in seen_pairs:
                            continue

                        orb = cls._angle_between(receiver.longitude, target_lon)
                        if orb <= vedic_orb:
                            aspect_label = f"Vedic-{caster.name}-{offset}th"
                            applying = cls._is_applying(caster, receiver, 0.0)
                            aspects.append(
                                AspectData(
                                    planet1=caster.name,
                                    planet2=receiver.name,
                                    aspect_type=aspect_label,
                                    exact_angle=round(target_lon, 4),
                                    orb=round(orb, 4),
                                    applying=applying,
                                    vedic=True,
                                )
                            )
                            seen_pairs.add(pair_key)

        # Sort by tightest orb first.
        aspects.sort(key=lambda a: a.orb)
        return aspects

File: lib/engines/aspect-engine.ts
// Planetary aspect calculation engine - port of backend/app/services/aspect_engine.py

import type { PlanetPosition } from "./swiss-ephemeris-engine";

// --------------------------------------------------------------------------
// Types
// --------------------------------------------------------------------------

export interface AspectData {
  planet1: string;
  planet2: string;
  aspect_type: string;
  exact_angle: number;
  orb: number;
  applying: boolean;
  vedic: boolean;
}

// --------------------------------------------------------------------------
// Constants
// --------------------------------------------------------------------------

const WESTERN_ASPECTS: Record<string, { angle: number; orb: number }> = {
  Conjunction: { angle: 0, orb: 8 },
  Opposition: { angle: 180, orb: 8 },
  Trine: { angle: 120, orb: 8 },
  Square: { angle: 90, orb: 7 },
  Sextile: { angle: 60, orb: 6 },
};

const VEDIC_SPECIAL_ASPECTS: Record<string, number[]> = {
  Mars: [4, 8],
  Jupiter: [5, 9],
  Saturn: [3, 10],
};

const PLANET_SPEED_ORDER: string[] = [
  "Moon", "Mercury", "Venus", "Sun", "Mars", "Jupiter", "Saturn", "Rahu", "Ketu",
];

// --------------------------------------------------------------------------
// Helpers
// --------------------------------------------------------------------------

function angleBetween(lon1: number, lon2: number): number {
  const diff = Math.abs(lon1 - lon2) % 360;
  return diff <= 180 ? diff : 360 - diff;
}

function isApplying(p1: PlanetPosition, p2: PlanetPosition, targetAngle: number): boolean {
  const speedRank: Record<string, number> = {};
  PLANET_SPEED_ORDER.forEach((name, idx) => {
    speedRank[name] = idx;
  });

  const rank1 = speedRank[p1.name] ?? 99;
  const rank2 = speedRank[p2.name] ?? 99;

  const faster = rank1 < rank2 ? p1 : p2;
  const slower = rank1 < rank2 ? p2 : p1;

  const currentAngle = angleBetween(faster.longitude, slower.longitude);
  return currentAngle >= targetAngle;
}

// --------------------------------------------------------------------------
// Main calculation
// --------------------------------------------------------------------------

export function calculateAspects(
  planets: PlanetPosition[],
  includeVedic = true
): AspectData[] {
  const aspects: AspectData[] = [];
  const seenPairs = new Set<string>();

  const n = planets.length;

  // Western aspects
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      const p1 = planets[i];
      const p2 = planets[j];
      const angle = angleBetween(p1.longitude, p2.longitude);

      for (const [aspectName, spec] of Object.entries(WESTERN_ASPECTS)) {
        const orb = Math.abs(angle - spec.angle);
        if (orb <= spec.orb) {
          const applying = isApplying(p1, p2, spec.angle);
          aspects.push({
            planet1: p1.name,
            planet2: p2.name,
            aspect_type: aspectName,
            exact_angle: spec.angle,
            orb: Math.round(orb * 10000) / 10000,
            applying,
            vedic: false,
          });
          const pairKey = [p1.name, p2.name].sort().join("|");
          seenPairs.add(pairKey);
        }
      }
    }
  }

  // Vedic special aspects (graha-drishti)
  if (includeVedic) {
    const vedicOrb = 10;

    for (const caster of planets) {
      const houseOffsets = VEDIC_SPECIAL_ASPECTS[caster.name];
      if (!houseOffsets) continue;

      for (const offset of houseOffsets) {
        const targetLon = (caster.longitude + offset * 30) % 360;

        for (const receiver of planets) {
          if (receiver.name === caster.name) continue;

          const pairKey = [caster.name, receiver.name].sort().join("|");
          if (seenPairs.has(pairKey)) continue;

          const orb = angleBetween(receiver.longitude, targetLon);
          if (orb <= vedicOrb) {
            const aspectLabel = `Vedic-${caster.name}-${offset}th`;
            const applying = isApplying(caster, receiver, 0);
            aspects.push({
              planet1: caster.name,
              planet2: receiver.name,
              aspect_type: aspectLabel,
              exact_angle: Math.round(targetLon * 10000) / 10000,
              orb: Math.round(orb * 10000) / 10000,
              applying,
              vedic: true,
            });
            seenPairs.add(pairKey);
          }
        }
      }
    }
  }

  // Sort by tightest orb
  aspects.sort((a, b) => a.orb - b.orb);
  return aspects;
}
